kernel/objects/rapl-driver-intel/objects/RaplDriverIntel.hh

```diff
@@ -0,0 +1,69 @@
+/* -*- mode:C++; indent-tabs-mode:nil; -*- */
+/* MIT License -- MyThOS: The Many-Threads Operating System
+ *
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use, copy,
+ * modify, merge, publish, distribute, sublicense, and/or sell copies
+ * of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * Copyright 2020 Philipp Gypser and contributors, BTU Cottbus-Senftenberg
+ */
+#pragma once
+
+#include "async/NestedMonitorDelegating.hh"
+#include "objects/IKernelObject.hh"
+
+namespace mythos {
+
+class RaplDriverIntel
+  : public IKernelObject
+{
+public:
+  optional<void const*> vcast(TypeId) const override { THROW(Error::TYPE_MISMATCH); }
+  optional<void> deleteCap(CapEntry&, Cap, IDeleter&) override { RETURN(Error::SUCCESS); }
+  void deleteObject(Tasklet*, IResult<void>*) override {}
+  void invoke(Tasklet* t, Cap self, IInvocation* msg) override;
+
+public:
+  RaplDriverIntel();
+  Error invoke_getRaplVal(Tasklet*, Cap, IInvocation* msg);
+  void printEnergy();
+
+protected:
+  /// @todo or one instance per hardware thread with HomeMonitor?
+  async::NestedMonitorDelegating monitor;
```

Comment:
Yes, the delegation will screw up the measurements. Let's assume that two cores do the system call simultaniously. The first one claims ownership of the RAPL kernel object and the second one gets enqueued in the delegation queue for later. Thus, the second core's measurement will be performed on the first core. Is that relevant for the RAPL measurements or would they receive the same information anyway because it's independent of cores?

Reply:
It should be core independent as long it is on the same socket. I prefer the measurement to be executed locally and synchronously. Hence, I will remove the delegation monitor.